src/main/kotlin/org/rust/ide/refactoring/move/common/RsMoveRetargetReferencesProcessor.kt

```diff
@@ -178,4 +186,64 @@ class RsMoveRetargetReferencesProcessor(
     }
+
+    // consider we move `foo1` and `foo2` from `mod1` to `mod2`
+    // this methods replaces `use mod1::{foo1, foo2};` to `use mod2::{foo1, foo2};`
+    //
+    // TODO: improve for complex use groups, e.g.:
+    //  `use mod1::{foo1, foo2::{inner1, inner2}};` (currently we create two imports)
+    private fun tryReplacePathOldInUseGroup(pathOld: RsPath, pathNew: RsPath): Boolean {
+        val useSpeck = pathOld.ancestorStrict<RsUseSpeck>() ?: return false
+        val useGroup = useSpeck.parent as? RsUseGroup ?: return false
+        val useItem = useGroup.ancestorStrict<RsUseItem>() ?: return false
+
+        // Before:
+        //   `use prefix::{mod1::foo::inner::{inner1, inner2}, other1, other2};`
+        //                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^ useGroup
+        //                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^ useSpeck
+        //                 ^~~~~~~~^ pathOld
+        // After:
+        //   `use prefix::{other1, other2};`
+        //   `use prefix::mod2::foo::inner::{inner1, inner2};`
+        pathOld.replace(pathNew)
+        val useSpeckText = useSpeck.text
+        // deletion should be before `insertUseItem`, otherwise `useSpeck` may be invalidated
+        useSpeck.deleteWithSurroundingComma()
+        if (useSpeckText.contains("::")) {
+            insertUseItemAndCopyAttributes(useSpeckText, useItem)
+        }
+
+        // TODO: group paths by RsUseItem and handle together ?
+        // consider pathOld == `foo1` in `use mod1::{foo1, foo2};`
+        // we just removed `foo1` from old use group and added new import: `use mod1::{foo2}; use mod2::foo1;`
+        // we can't optimize use speck right now,
+        // because otherwise `use mod1::{foo2};` becomes `use mod1::foo2;` which destroys `foo2` reference
+        useSpecksToOptimize += useGroup.parentUseSpeck
+        return true
+    }
+
+    private fun insertUseItemAndCopyAttributes(useSpeckText: String, existingUseItem: RsUseItem) {
+        val containingMod = existingUseItem.containingMod
+        if (existingUseItem.outerAttrList.isEmpty() && existingUseItem.vis == null) {
+            containingMod.insertUseItem(psiFactory, useSpeckText)
+        } else {
+            val useItem = existingUseItem.copy() as RsUseItem
+            val useSpeck = psiFactory.createUseSpeck(useSpeckText)
+            useItem.useSpeck!!.replace(useSpeck)
+            containingMod.addAfter(useItem, existingUseItem)
+        }
+        filesToOptimizeImports.add(containingMod.containingFile as RsFile)
+    }
+
+    private fun addImport(context: RsElement, usePath: String) {
+        addImport(psiFactory, context, usePath)
+        filesToOptimizeImports.add(context.containingFile as RsFile)
+    }
+
+    fun optimizeImports() {
+        useSpecksToOptimize.forEach { RsImportOptimizer.optimizeUseSpeck(psiFactory, it) }
+
+        filesToOptimizeImports.addAll(useSpecksToOptimize.mapNotNull { it.containingFile as? RsFile })
```

Comment:
I'm not sure why do you use safe case here and unsafe cast in other places (`context.containingFile as RsFile`, `containingMod.containingFile as RsFile`, etc.)? Is this case different from the others?

Reply:
Thanks, fixed (it was different because `containingFile` can be `DummyHolder`)